src/stash/src/lib.rs

```diff
@@ -1000,6 +1012,45 @@ where
         }
     }
 
+    /// Set the value for a key. Returns the previous entry if the key existed,
+    /// otherwise None.
+    ///
+    /// Returns an error if the uniqueness check failed.
+    pub fn set(&mut self, k: K, v: Option<V>) -> Result<Option<V>, StashError> {
+        // Save the pending value for the key so we can restore it in case of
+        // uniqueness violation.
+        let restore = self.pending.get(&k).cloned();
+
+        let prev = match self.pending.entry(k.clone()) {
+            // key hasn't been set in this txn yet. Set it and return the
+            // initial txn's value of k.
+            Entry::Vacant(e) => {
+                e.insert(v);
+                self.initial.get(&k).cloned()
+            }
+            // key has been set in this txn. Set it and return the previous
+            // pending value.
+            Entry::Occupied(mut e) => e.insert(v),
+        };
+
+        // Check for uniqueness violation.
+        if let Err(err) = self.verify() {
+            // Revert self.pending to the state it was in before calling this
+            // function.
+            match restore {
+                Some(v) => {
+                    self.pending.insert(k, v);
+                }
+                None => {
+                    self.pending.remove(&k);
+                }
+            }
+            Err(err)
+        } else {
+            Ok(prev)
+        }
```

Comment:
I have a couple thoughts about the uniqueness violations. None of them need to be implemented in this PR, but it just got me thinking about them.
- A lot of the checks are just no-ops that return false. Maybe we should change the type of the callback to `Option<fn(a: &V, b: &V) -> bool>` and only iterate through collections that have a valid callback.
- Maybe we want to move this call to only happen on transaction commit. For each collection that has a non-empty `pending` map, we call `verify`. That way we only have to iterate through everything once per transaction.

Reply:
+1 on changing the verify method to not iterate if it doesn't need to.

> only happen on transaction commit

That sounds a bit like the `DEFERRED` constraint from https://www.postgresql.org/docs/current/sql-set-constraints.html, and I think won't prevent violations mid transaction that are resolved before the transaction ends. I think that'd be fine as long as we audit all uses of TableTransaction to make sure that behavior is acceptable to them.